core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroup.java

```diff
@@ -187,25 +190,37 @@ private void refreshContexts() {
                     }
 
                     removed = true;
+                    willRemove.add(e.getValue());
                     i.remove();
-                    e.getValue().destroy();
                 }
 
                 if (newSelectedEndpoints.isEmpty()) {
                     if (!removed) {
                         // The weight of an endpoint is changed. So we just refresh.
                         refreshEndpoints();
+                    } else {
+                        willRemove.forEach(DefaultHealthCheckerContext::destroy);
                     }
                     return;
                 }
 
                 // At this time newSelectedEndpoints only contains newly appeared endpoints.
                 // Start the health checkers with new contexts.
+                final List<CompletableFuture<?>> initialFutures = new ArrayList<>(newSelectedEndpoints.size());
                 for (Endpoint e : newSelectedEndpoints) {
                     final DefaultHealthCheckerContext ctx = new DefaultHealthCheckerContext(e);
                     ctx.init(checkerFactory.apply(ctx));
+                    initialFutures.add(ctx.initialCheckFuture);
                     contexts.put(e, ctx);
                 }
+                if (!willRemove.isEmpty()) {
+                    // Remove old endpoints after finish initial health check of `newSelectedEndpoints`.
+                    CompletableFutures.allAsList(initialFutures)
```

Comment:
`CompletableFutures.allAsList()` will not wait for all futures to be completed if there is an error between the futures.
How about using `CompletableFuture.allOf(initialFutures)` in order to wait for all futures.

Reply:
I choosed allAsList because I thought, waiting at least one healthcheck would be enough but do you think, better to wait all futures?

Reply:
I thought of a worst-case that if a broken server returns 503 Service Unavailable quickly, the old endpoints can be destroyed before receiving a healthy response from new endpoints. Then, the endpoint group could be empty at the time.

Reply:
> waiting at least one healthcheck 

If it is enough to wait for the first successful result, don't we need a custom logic that will complete a future when receiving the first success?

Reply:
>  a worst-case that if a broken server returns 503 Service Unavailable quickly

Ah, understood. Let me wait all the futures then. (and add option to wait only first Future later if really needed)